### DISCUSSION TITLE
Why do the docs state that the Mastodon frontend is in React?

### DISCUSSION THREAD

**Discussion by crawlchange** [mastodon/mastodon#19336] [General]
Looking at Mastodon code, it is clearly using haml, not React. I could understand if it was a situation were haml delegated to React at some point. But nope. Only haml. The React components exist, but are never called. One example would be the "Status" component, which is never called under the statuses views.

The only mention to this in the documentation is regarding the directories, where it states:

app/views -> "Templates for generating HTML or other output"
app/javascript/mastodon -> "Code for the multi-column React.js application"

What is the "multi-column React.js application"? Where is anything said about that "multi-column React.js application"? Where do I download it? How do I run it?

What practices are being followed in this "React in the frontend" architecture? Where can a new developer be referred to? React-rails gems are not used. Standard rails practices are not being followed.

As you can see documentation is severely lacking on some of the most crucial aspects of the mastodon architecture. It makes almost impossible for a new developer to come in without spending a huge amount of time reverse-engineering an undocumented code architecture.

**Comment by Gargron** [comment]
To enable the multi-column layout, opt into the "advanced web interface" in your appearance preferences. By default the layout is not multi-column anymore, so that reference is a bit outdated. You can see the React frontend when you are logged in. In the main branch, you also see it when you are logged out.

**Comment by crawlchange** [comment]
Is React only used for the multi-column layout? Without the "multi-column layout" enabled, the frontend uses no javascript?

**Reply by trwnh** [reply]
No, the React code handles the entire web-app. The description ("Code for the multi-column React.js application") was written before the single-column view was implemented.

**Comment by crawlchange** [comment]
"You can see the React frontend when you are logged in."

How do I see it? And even if I do, how does it get there? Where is react being called in the code? Is it a single page application? How does the "react frontend" fetch data from the backend?

As a specific example to make matters simple, where and how in the code does Rails mount and pass data into the Status react component?

**Reply by trwnh** [reply]
The Rails code is used to serve a REST API which the React frontend consumes

**Comment by trwnh** [comment]
> What practices are being followed in this "React in the frontend" architecture? Where can a new developer be referred to? React-rails gems are not used. Standard rails practices are not being followed.

In short:

- the backend is a Ruby on Rails application that powers a REST API. HAML is used only for the settings pages and for the public-facing pages (status and profile permalinks). In the latest code on the `main` branch, work is currently being done to deprecate the HAML-based status and post permalinks.
- the frontend is a React application that consumes the REST API. This frontend is used for the entire web application, except for the settings pages. It is built with Webpacker.

As far as any developer is concerned, these are two separate applications. There is no coupling between the frontend and the backend.

**Comment by crawlchange** [comment]
"work is currently being done to deprecate the HAML-based status and post permalinks."

Nice!

So how is the React frontend served? There is no gem installed for react. Is it a SPA? Where is the code called? Which controller returns or calls the react code? What architecture is being followed, where can a developer be referred to understand this architecture?

**Reply by trwnh** [reply]
I believe it is an SPA, yes. There is no gem (nor is one needed) because it is all compiled via Webpacker.

The relevant HAML I could find is https://github.com/mastodon/mastodon/blob/main/app/views/home/index.html.haml

```haml
- content_for :header_tags do
  = render partial: 'shared/og'

= render 'shared/web_app'
```

This renders from https://github.com/mastodon/mastodon/blob/main/app/views/shared -- the header tags contain the OpenGraph preview information, and the web app is mounted after that inside the body.

**Reply by crawlchange** [reply]
Thanks, I guess that's a good starting point.
